pkg/controllers/clusterca/clusterca_controller.go

```diff
@@ -56,50 +56,82 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
 
-	if err != nil {
-		return ctrl.Result{}, err
+	if needUpdateMCI {
+		clusterInfo.Status.DistributionInfo.OCP.LastAppliedAPIServerURL =
+			clusterInfo.Status.DistributionInfo.OCP.ManagedClusterClientConfig.URL
+		// retry if update managed cluster info status failed with conflict, because if we do not
+		// retry but return err here, next reconcile will stop at "updateClientConfig", since the
+		// clientConfig of the ManagedCluster CR has already been updated successfully above in the current
+		// reconcile, the "updateClientConfig" func will return needUpdateMCI=false in the next reconcile loop.
+		err = r.UpdateManagedClusterInfoStatus(ctx, clusterInfo)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
-// updateClientConfig merge config from clusterinfoconfigs to clusterconfigs
-func updateClientConfig(clusterConfigs []clusterv1.ClientConfig, clusterinfoConfig clusterinfov1beta1.ClientConfig) ([]clusterv1.ClientConfig, bool) {
+func (r *ClusterCaReconciler) UpdateManagedClusterInfoStatus(
+	ctx context.Context, instance *clusterinfov1beta1.ManagedClusterInfo) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		return r.client.Status().Update(ctx, instance)
+	})
+}
+
+// updateClientConfig merge config from clusterinfoconfigs to clusterconfigs, it returns 3 parameters:
+//   - the updated cluster clientConfig
+//   - whether it is needed to update the managedCluster object
+//   - whether it is needed to update the managedClusterInfo object, only if the lastAppliedURL
+//     changes, this will be true
+func updateClientConfig(clusterConfigs []clusterv1.ClientConfig, clusterinfoConfig clusterinfov1beta1.ClientConfig,
+	lastAppliedURL string) ([]clusterv1.ClientConfig, bool, bool) {
 	// If clusterinfo config is null return clusterconfigs
 	if len(clusterinfoConfig.URL) == 0 {
-		return clusterConfigs, false
+		return clusterConfigs, false, false
 	}
 
+	// The lastAppliedURL comes from the value of infrastructures config(status.apiServerURL), if the
+	// infrastructures config value changes, here we will replace the old clientConfig URL instead of
+	// prepending a new one.
 	for i, cluConfig := range clusterConfigs {
-		if cluConfig.URL != clusterinfoConfig.URL {
+		if cluConfig.URL != clusterinfoConfig.URL && cluConfig.URL != lastAppliedURL {
 			continue
 		}
-		if !reflect.DeepEqual(cluConfig.CABundle, clusterinfoConfig.CABundle) {
-			clusterConfigs[i].CABundle = clusterinfoConfig.CABundle
-			return clusterConfigs, true
+
+		if reflect.DeepEqual(cluConfig.CABundle, clusterinfoConfig.CABundle) && cluConfig.URL == clusterinfoConfig.URL {
+			return clusterConfigs, false, false
 		}
-		return clusterConfigs, false
+
+		clusterConfigs[i].URL = clusterinfoConfig.URL
+		clusterConfigs[i].CABundle = clusterinfoConfig.CABundle
+		return clusterConfigs, true, cluConfig.URL != clusterinfoConfig.URL
```

Comment:
cluConfig.URL != clusterinfoConfig.URL this seems buggy, what happens if updateMC succeed and updateMCI fails? I think you should always check if lastApplied equals clusterinfoConfig after updateMC succeeds

Reply:
fixed, will always check if lastApplied equals clusterinfoConfig no matter  managedCluster needs to update or not.